### PR TITLE
Add a kwarg for num_processes in the transpiler

### LINKED ISSUE #9365: Expose process-count limits on a per-`transpile` basis
### What should we add?

It's currently possible to limit the number of processes used by all calls to `transpile` in a given Python process by [the configuration file entry `num_processes`](https://qiskit.org/documentation/configuration.html), or by the environment variable `QISKIT_NUM_PROCS`, but both of these are only read on `import qiskit`, and affect every call to `transpile`.

It would be good to expose a `num_processes` kwarg in `transpile`, which overrides either of these two options, and can be used to restrict an individual call to certain behaviour.  We already have the right kwarg available in `parallel.parallel_map`, so exposing this to users should mostly be as straightforward as adding an option to `transpile` and propagating it through.

We might want to re-arrange the internal logic of `parallel_map` a little so that it's easier to pass `num_processes=None` without clobbering its internal default of taking the value from the environment if the option is empty.

### PR DESCRIPTION
**Pull request by Ca1eb3**
### Summary
fixes #9365 
Add a kwarg to the transpiler which allows for overriding the user configs number of processes. As per this comment https://github.com/Qiskit/qiskit/issues/9365#issuecomment-1881205239 the order in which number of processes should be determined is the following:
1. Keyword arguments given explicitly
2. An argument set by environment variable (i.e. QISKIT_NUM_PROCS in this case)
3. An argument read from a user's configuration file
4. The system default

### Details and comments
This required modifications to the transpiler to serialize all of the instruction durations before running any of the circuits. I am unsure if this is the best way to go about serialization of the instruction durations and would welcome suggestions of a more straight forward way to achieve this. This also requires modifications to the parallel_map function to allow for None to be passed in as the number of processes and to take on the default specified by the user config or the environment variable.

### CODE REVIEW

File: qiskit/compiler/transpiler.py [hunk 1]
```diff
@@ -257,6 +259,11 @@ def callback_func(**kwargs):
             to support custom compilation target-specific passes/plugins which support
             backend-specific compilation techniques. If you'd prefer that these defaults were
             not used this option is used to disable those backend-specific defaults.
+        num_processes: The maximum number of parallel processes to launch for this call to
+            transpile if parallel execution is enabled. This argument overrides
+            num_processes in the user configuration file, and the QISKIT_NUM_PROCS
```
Comment (mtreinish) [inline qiskit/compiler/transpiler.py]:
```suggestion
            the ``num_processes`` field in the user configuration file, and the ``QISKIT_NUM_PROCS``
```

File: qiskit/compiler/transpiler.py [hunk 2]
```diff
@@ -257,6 +259,11 @@ def callback_func(**kwargs):
...
+            environment variable. If set to None the system default or local user configuration
```
Comment (mtreinish) [inline qiskit/compiler/transpiler.py]:
```suggestion
            environment variable. If set to ``None`` the system default or local user configuration
```

File: releasenotes/notes/add-num-processes-kwarg-to-transpiler-3cb7f3457b54a535.yaml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+---
+features:
+  - |
+    Add new keyword argument num_processes to :func:`~qiskit.compiler.transpiler`.
```
Comment (mtreinish) [inline releasenotes/notes/add-num-processes-kwarg-to-transpiler-3cb7f3457b54a535.yaml]:
```suggestion
    Add new keyword argument ``num_processes`` to :func:`~qiskit.compiler.transpiler`.
```

File: qiskit/compiler/transpiler.py [hunk 4]
```diff
@@ -356,76 +363,52 @@ def callback_func(**kwargs):
...
+            pair = (circuits[i], instruction_duration)
+            circuits_instruction_durations.append(pair)
+    else:
+        for i in range(0, len(circuits)):
+            pair = (circuits[i], None)
+            circuits_instruction_durations.append(pair)
+
+    out_circuits = []
+    out_circuits = parallel_map(
+        _run_circuits,
```
Comment (mtreinish) [inline qiskit/compiler/transpiler.py]: This is potentially problematic as we're moving the unit of parallelism from being `PassManager.run` to the outer function that builds the pass manager and then runs it in a parallel context. This is actually how things used to be setup in `transpile` before and we moved to building a single passmanager and doing the parallel dispatch inside `.run()` both for consistency in the execution paths and also ebcause it's lower overhead for most multiprocessing modes. Using `PassManager.run()` avoids a bunch of serialization and IPC as `task_args` have to be directly serialized and sent via IPC to the child processes.

**Review by mtreinish — COMMENTED** [review]
Thanks for opening this PR, I'm definitely in favor of adding this argument and this is a good start. I have some small concerns about the current implementation that I left an inline comment on. I think that we should be adding an equivalent `num_processes` kwarg to the `PassManager.run` method (which internally can call `parallel_map`) definition and leverage that from inside `transpile`.

File: qiskit/compiler/transpiler.py (hunk 4 above)
Comment (Ca1eb3) [reply qiskit/compiler/transpiler.py]: Ok that makes sense I will update accordingly.

Comment (Ca1eb3) [reply qiskit/compiler/transpiler.py]: Changes have been made to pass the num_processes through PassManager.run

**Review by mtreinish — APPROVED** [review]
This LGTM now, thanks for the quick update

**Pull request merged**